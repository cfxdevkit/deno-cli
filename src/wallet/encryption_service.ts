import { crypto } from '@std/crypto'
import { Secret } from 'cliffy/prompt'

export class EncryptionService {
	async deriveKeyFromPassword(reason: string, salt: Uint8Array): Promise<CryptoKey> {
		const password = await Secret.prompt({ message: reason })
		const encoder = new TextEncoder()
		const passwordKey = await crypto.subtle.importKey(
			'raw',
			encoder.encode(password),
			'PBKDF2',
			false,
			['deriveKey'],
		)

		return crypto.subtle.deriveKey(
			{
				name: 'PBKDF2',
				salt,
				iterations: 100_000,
				hash: 'SHA-256',
			},
			passwordKey,
			{ name: 'AES-GCM', length: 256 },
			true,
			['encrypt', 'decrypt'],
		)
	}

	getRandomValues(array: Uint8Array) {
		return crypto.getRandomValues(array)
	}

	async encryptMnemonic(mnemonic: string): Promise<string> {
		const iv = this.getRandomValues(new Uint8Array(12))
		const salt = this.getRandomValues(new Uint8Array(16))
		const key = await this.deriveKeyFromPassword('Enter encryption password to secure your mnemonic.', salt)
		const encrypted = await crypto.subtle.encrypt(
			{ name: 'AES-GCM', iv },
			key,
			new TextEncoder().encode(mnemonic),
		)
		const encryptedData = new Uint8Array([...salt, ...iv, ...new Uint8Array(encrypted)])
		return btoa(String.fromCharCode(...encryptedData))
	}

	async decryptMnemonic(encryptedMnemonic: string): Promise<string> {
		const encryptedBytes = Uint8Array.from(atob(encryptedMnemonic), (c) => c.charCodeAt(0))
		const salt = encryptedBytes.slice(0, 16)
		const iv = encryptedBytes.slice(16, 28)
		const data = encryptedBytes.slice(28)

		for (let attempt = 1; attempt <= 3; attempt++) {
			try {
				const key = await this.deriveKeyFromPassword('Password: ', salt)
				const decrypted = await crypto.subtle.decrypt({ name: 'AES-GCM', iv }, key, data)
				return new TextDecoder().decode(decrypted)
			} catch {
				console.error(`Decryption failed (${attempt}/3). Incorrect password or corrupted data.`)
			}
		}
		throw new Error('Maximum decryption attempts reached.')
	}
}
