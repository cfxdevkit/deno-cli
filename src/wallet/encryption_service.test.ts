import { snapshotTest } from 'cliffy/testing'
import { ansi } from 'cliffy/ansi'
import { EncryptionService } from './encryption_service.ts'

import { assertEquals } from '@std/assert'
Deno.test('EncryptionService - encrypt and decrypt mnemonic successfully', async () => {
	await snapshotTest({
		name: 'EncryptionService - encrypt and decrypt mnemonic successfully',
		meta: import.meta,
		stdin: ansi
			.text('password\n')
			.text('password\n')
			.toArray(),
		async fn() {
			const encryptionService = new EncryptionService()
			const mnemonic = 'test test test test test test test test test test test junk'
			encryptionService.getRandomValues = (typedArray: Uint8Array) => {
				return typedArray
			}
			const encryptedMnemonic = await encryptionService.encryptMnemonic(mnemonic)

			assertEquals(
				encryptedMnemonic,
				'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAANtJQA7N38hVEuTdGw+i+TKMDYWq4lJ7+eVgQKvL3RPe8wniDZH8IBhbyX/BlfgviIBIOHEAswZdqpCJUruE5Nda3rF8xMumtK6jjg==',
			)
			assertEquals(mnemonic, await encryptionService.decryptMnemonic(encryptedMnemonic))
		},
	})
})
