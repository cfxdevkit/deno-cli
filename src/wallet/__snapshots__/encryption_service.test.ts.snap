export const snapshot = {};

snapshot[`EncryptionService - encrypt and decrypt mnemonic successfully 1`] = `
stdout:
"? Enter encryption password to secure your mnemonic. › \\x1b[56G\\x1b[G\\x1b[0J
? Enter encryption password to secure your mnemonic. › ********\\x1b[64G\\x1b[G\\x1b[0J
? Enter encryption password to secure your mnemonic. › ********
\\x1b[?25h\\x1b[?25h? Password:  › \\x1b[16G\\x1b[G\\x1b[0J
? Password:  › ********\\x1b[24G\\x1b[G\\x1b[0J
? Password:  › ********
\\x1b[?25h\\x1b[?25h"
stderr:
""
`;
